README.md

```diff
@@ -1,2 +1,313 @@
+            REPLICATELOG(nodeld, follower)
+        end for
+    else
+        forward the request to currentLeader via a FIFO link
+    end if
+end on
+
+periodically at node nodeld do
+    if currentRole = leader then
+        for each follower in nodes \ {nodeld} do
+            REPLICATELOG(nodeld, follower)
+        end for
+    end if
+end on
+
+```
+
+This phase handles broadcasting messages in the cluster.
+
+* If the node is the leader, it appends the message to its log and replicates it to the followers.
+* If the node is not the leader, it forwards the request to the current leader.
+
+## Replicating from Leader to Followers
+
+```shell
+function REPLICATELOG(leaderId, followerId)
+    prefixLen := sentLength[followerId]
+    suffix := (log[prefixLen+1:], ...)
+    prefixTerm := 0
+    if prefixLen > 0 then
+        prefixTerm := log[prefixLen].term
+    end if
+    send (LogRequest, leaderId, currentTerm, prefixLen, prefixTerm, commitLength, suffix) to followerId
+end function
+
+```
+This phase handles log replication from the leader to its followers.
+
+* The `REPLICATELOG` function iscalled when there are new log entries or periodically.
+* The function determines the prefix and suffix of the log entries to send to a follower.
+* A `LogRequest` message is sent to the follower containing the necessary log information.
+
+## Followers Receiving Messages
+
+```shell
+on receiving (LogRequest, leaderId, term, prefixLen, prefixTerm, leaderCommit, suffix) at node nodeld do
+    if term > currentTerm then
+        currentTerm := term
+        currentRole := follower
+        currentLeader := leaderId
+    end if
+    logOk := (log.length >= prefixLen) and (prefixLen = 0 or log[prefixLen].term = prefixTerm)
+    if term = currentTerm and logOk then
+        APPENDENTRIES(prefixLen, leaderCommit, suffix)
+        ack := prefixLen + suffix.length
+        send (LogResponse, nodeld, currentTerm, ack, true) to leaderId
+    else
+        send (LogResponse, nodeld, currentTerm, 0, false) to leaderId
+    end if
+end on
+
+```
+
+This phase handles the log replication process from the leader to its followers.
+
+* If the received term is greater than the current term, the node updates its term and role.
+* The follower checks the consistency of the received log with its own log.
+* If the logs are consistent, the follower appends the new entries and responds with an acknowledgment.
+* If the logs are inconsistent, the follower responds with a negative acknowledgment.
+
+## Updating Followers' Logs
+
+```shell
+function APPENDENTRIES(prefixLen, leaderCommit, suffix)
+    if suffix.length > 0 and log.length >= prefixLen then
+        index := min(log.length, prefixLen + suffix.length) - 1
+        if log[index].term != suffix[index - prefixLen].term then
+            log := log[1:prefixLen]
+        end if
+    end if
+    if prefixLen + suffix.length > log.length then
+        for i := log.length - prefixLen + 1 to suffix.length do
```

Comment:
Instead of "i := log.length - prefixLen + 1 to suffix.length", it should be "i := log.length - prefixLen to suffix.length - 1"